### PR TITLE
chore: remove PyTorch 2.5.0 checks

### PR DESCRIPTION
**Pull request by JP-sDEV**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation

#1861 

#### Changelog
What are the changes made in this PR?
* modify documentation and checks for PyTorch 2.5.0

Note: Line 69 in `torchtune/torchtune/training/_activation_offloading.py`, should this `if` block be removed? or modified to check for pytorch `2.5.0`? - was not included in the issue to modify this block.
```
 if use_streams is None:
            # Default to True if an acceptable torch is installed (later nightly/version or from source)
            **self.use_streams = torch.__version__ >= "2.5.0.dev20240907"** 
```

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: docs/source/tutorials/memory_optimizations.rst
Comment (RdoubleA) [inline docs/source/tutorials/memory_optimizations.rst:111]:
```suggestion
usage of streams, make sure you are on a torch version equal to or later than PyTorch 2.5.0.
```

File: torchtune/training/_activation_offloading.py
Comment (RdoubleA) [inline torchtune/training/_activation_offloading.py:36]:
```suggestion
            after torch-2.5.0. Default: True.
```

File: torchtune/training/_compile.py
Comment (RdoubleA) [inline torchtune/training/_compile.py:56]: I'm not sure if we want to retain the fallback logic for older pytorch versions. If so, then the if-else should remain the same and only the warning message should be updated. any thoughts? @ebsmothers @felipemello1

**Review by RdoubleA — COMMENTED** [review]
Thanks again for doing these - looks great. A few minor nits and one comment we should decide on.

**Comment by JP-sDEV** [comment]
> Thanks again for doing these - looks great. A few minor nits and one comment we should decide on.

Thanks, will keep up with the updates!  Once finalized, let me know what updates are needed. Cheers! :)

File: torchtune/training/_compile.py
Comment (ebsmothers) [reply torchtune/training/_compile.py:56]: Sorry just seeing this now. I think if we claim to not support PyTorch < 2.5 then we shouldn't leave in the full-model compile option at all. For the same reason I'm ambivalent about leaving in the log warning.. really if we want to check someone is at least on the latest stable PyTorch we should just do it in a single consolidated place. So not the end of the world to keep the warning in, but personally I'd just take it out.

Comment (ebsmothers) [reply torchtune/training/_compile.py:56]: I vote remove it altogether. We claim to only support latest stable PyTorch so no need to keep this around. We can treat adding a consolidated PyTorch version check as a separate task

**Comment by RdoubleA** [comment]
Hey @JP-sDEV, pinging this thread again to see if you're able to resolve the merge conflicts so we can get this landed. I believe we agreed that we can remove the fallback logic in `_compile.py`, so would be great to make that change as well.

**Comment by JP-sDEV** [comment]
Hi @RdoubleA, apologies for the delay - will resolve the conflicts and remove the fallback logic first thing tomorrow morning (EST)

File: docs/source/tutorials/memory_optimizations.rst
Comment (RdoubleA) [reply docs/source/tutorials/memory_optimizations.rst:113]: bumping this

File: torchtune/training/_activation_offloading.py
Comment (RdoubleA) [inline torchtune/training/_activation_offloading.py:73]: let's remove the check below and make `use_streams: bool = True`

Comment (JP-sDEV) [reply torchtune/training/_activation_offloading.py:73]: are you referring to this `use_streams` check found in `__init__`? 

```python
        if use_streams is None:
            # Default to True if an acceptable torch is installed (later nightly/version or from source)
            self.use_streams = torch.__version__ >= "2.5.0.dev20240907"
        else:
            self.use_streams = use_streams
```

File: torchtune/modules/attention_utils.py
Comment (RdoubleA) [reply torchtune/modules/attention_utils.py:115]: the `_attention_call` is modified to support both flex and sdpa if `_SUPPORTS_FLEX_ATTENTION` on module initialization, but it still needs the capability to fallback to SDPA if a block mask is not passed during runtime. So this awkward logic should be retained

Comment (RdoubleA) [reply torchtune/modules/attention_utils.py:115]: > We should always be able to import BlockMask on stable, so we could move [this import](https://github.com/pytorch/torchtune/blob/4b6877a6ef31a1f987c27594eaf8fe467b5ab785/torchtune/modules/attention_utils.py#L20) out of the flex attention support check.

we'd still need to verify their hardware can run it

File: torchtune/training/_activation_offloading.py
Comment (JP-sDEV) [reply torchtune/training/_activation_offloading.py:73]: I have noticed that `OffloadingActivations`  in `torchtune/training/_activation_offloading.py` still checks for torch version: `2.5.0.dev20240907`. Should this check also be removed? 

```python
        # for streaming
        if self.use_streams:
            if torch.__version__ < "2.5.0.dev20240907":
                raise RuntimeError(
                    "OffloadActivations with use_streams=True requires PyTorch 2.5.0.dev20240907 or later."
                )
```

Comment (RdoubleA) [reply torchtune/training/_activation_offloading.py:73]: Yes, good catch. Let's remove that as well. I believe these may have been added after I put the issue up, or maybe I just missed it

Comment (JP-sDEV) [reply torchtune/training/_activation_offloading.py:73]: sounds good, will update all `2.5.0.dev20240907` checks in the file

File: torchtune/modules/attention_utils.py
Comment (SalmanMohammadi) [reply torchtune/modules/attention_utils.py:115]: rightio, because we're returning a callable the check needs to happen outside. thanks for explaining!

**Comment by SalmanMohammadi** [comment]
Can we update these checks here? Are we able to remove them?
https://github.com/pytorch/torchtune/blob/4b6877a6ef31a1f987c27594eaf8fe467b5ab785/recipes/qat_distributed.py#L139
https://github.com/pytorch/torchtune/blob/4b6877a6ef31a1f987c27594eaf8fe467b5ab785/recipes/full_finetune_distributed.py#L128

File: docs/source/tutorials/memory_optimizations.rst
Comment (SalmanMohammadi) [reply docs/source/tutorials/memory_optimizations.rst:113]: I believe we can just revert the changes here to leave the file as-as since it's been updated in another PR @JP-sDEV

File: tests/torchtune/modules/test_attention_utils.py
Comment (RdoubleA) [inline tests/torchtune/modules/test_attention_utils.py:85]: Ok looks like we need to keep this check, in case the hardware that runs the gpu tests on GitHub CI does not support flex attention

**Comment by RdoubleA** [comment]
@JP-sDEV thanks for pushing this through, just one more comment which should fix the failing gpu test

**Comment by JP-sDEV** [comment]
> @JP-sDEV thanks for pushing this through, just one more comment which should fix the failing gpu test

I am reading the comments and reviewing the commits, and if I understand correctly - the changes needed for checking hardware compatibility is something outside this issue? 

If so, I'll merge with the main branch once resolved :)

**Comment by RdoubleA** [comment]
The update is that we just need to keep the decorator that was removed on `test_packed_block_causalmask_flex`, then it should be good to go

**Review by RdoubleA — APPROVED** [review]

**Comment by RdoubleA** [comment]
@JP-sDEV I went ahead and quickly made the change - thank you again for your help!

**Comment by JP-sDEV** [comment]
No problem! Thank you for guiding me through the issue and for being patient! :)

**Pull request merged**
